src/main/java/com/spikes2212/command/drivetrains/swerve/SwerveDrivetrain.java

```diff
@@ -0,0 +1,231 @@
+     * @param backLeftModule            the back left module using {@link SwerveModule}
+     * @param backRightModule           the back right module using{@link SwerveModule}
+     * @param halfDrivetrainTrackWidth  the width of the drivetrain
+     * @param halfDrivetrainTrackLength the length of the drivetrain
+     * @param maxPossibleVelocity       the maximum possible velocity of the drive motors
+     */
+    public SwerveDrivetrain(SwerveModule frontLeftModule, SwerveModule frontRightModule, SwerveModule backLeftModule,
+                            SwerveModule backRightModule, double halfDrivetrainTrackWidth, double halfDrivetrainTrackLength,
+                            double maxPossibleVelocity) {
+        this(getClassName(DEFAULT_NAMESPACE_NAME), frontLeftModule, frontRightModule, backLeftModule, backRightModule,
+                halfDrivetrainTrackWidth, halfDrivetrainTrackLength, maxPossibleVelocity);
+    }
+
+    /**
+     * Function that moves the robot in a swerve motion.
+     *
+     * @param xSpeed          the desired speed on the x-axis.
+     * @param ySpeed          the desired speed on the y-axis.
+     * @param rotationSpeed   the desired speed for the drivetrain rotation.
+     * @param isFieldRelative whether the drive should be relative to the field or to the robot.
+     * @param timeStep        the derivation of time  the speed should be applied.
+     * @param usePIDVelocity  whether the robot velocity will be applied using P.I.D or not.
+     */
+    public void drive(double xSpeed, double ySpeed, double rotationSpeed, boolean isFieldRelative,
+                      double timeStep, boolean usePIDVelocity) {
+        SwerveModuleState[] states;
+        if (isFieldRelative) {
+            states = getFieldRelativeSpeeds(xSpeed, ySpeed, rotationSpeed, timeStep);
+        } else {
+            states = getRobotRelativeSpeeds(xSpeed, ySpeed, rotationSpeed, timeStep);
+        }
+        SwerveDriveKinematics.desaturateWheelSpeeds(states, maxPossibleVelocity);
+        setTargetModuleStates(states, usePIDVelocity);
+    }
+
+    /**
+     * Sets the desired module states.
+     *
+     * @param targetModuleStates an array of the desired module states
+     * @param usePIDVelocity     whether the modules will drive with P.I.D for the velocity
+     */
+    protected void setTargetModuleStates(SwerveModuleState[] targetModuleStates, boolean usePIDVelocity) {
+        frontLeftModule.setTargetState(targetModuleStates[0], maxPossibleVelocity, usePIDVelocity);
+        frontRightModule.setTargetState(targetModuleStates[1], maxPossibleVelocity, usePIDVelocity);
+        backLeftModule.setTargetState(targetModuleStates[2], maxPossibleVelocity, usePIDVelocity);
+        backRightModule.setTargetState(targetModuleStates[3], maxPossibleVelocity, usePIDVelocity);
+    }
+
+    /**
+     * Allows updating the robot angle using an external sensor.
```

Comment:
I would also specify for the sake of library documentation that this resets the field relativity to be same as robot relativity